### ISSUE TITLE
Nested dataclass definitions not working with pydantic

### ISSUE THREAD

**Issue by and3rson** [pydantic/pydantic#4877]
### Initial Checks

- [X] I have searched GitHub for a duplicate issue and I'm sure this is something new
- [X] I have searched Google & StackOverflow for a solution and couldn't find anything
- [X] I have read and followed [the docs](https://pydantic-docs.helpmanual.io) and still think this is a bug
- [X] I am confident that the issue is with pydantic (not my code, or another library in the ecosystem like [FastAPI](https://fastapi.tiangolo.com) or [mypy](https://mypy.readthedocs.io/en/stable))

### Description

When defining dataclasses inside dataclasses, and then using those top-level dataclasses in pydantic models, I'm sometimes getting a `TypeError: non-default argument '...' follows default argument`.

Removing definitions of child dataclasses makes the code work properly.

### Example Code

```Python
from typing import Optional, List
from dataclasses import dataclass, field

from pydantic import BaseModel


@dataclass
class Foo:
    a: Optional[int] = field(default=42)
    b: List = field(default_factory=list)

    @dataclass
    class Bar:
        pass


@dataclass
class Model1:
    foo: Foo
print(Model1(foo=Foo()))  # Model(foo=Foo(a=42, b=[]))


class Model2(BaseModel):
    foo: Foo
print(Model2(foo=Foo()))  # TypeError: non-default argument 'b' follows default argument
# Removing the definition of `Foo.Bar` makes this work properly.
```

### Python, Pydantic & OS Version

```Text
pydantic version: 1.10.2
            pydantic compiled: True
                 install path: /home/anderson/src/***/.venv/lib/python3.9/site-packages/pydantic
               python version: 3.9.13 (main, Oct  5 2022, 17:32:54)  [GCC 12.2.0]
                     platform: Linux-6.0.9-arch1-1-x86_64-with-glibc2.36
     optional deps. installed: ['typing-extensions']
```

### Affected Components

- [ ] [Compatibility between releases](https://pydantic-docs.helpmanual.io/changelog/)
- [ ] [Data validation/parsing](https://pydantic-docs.helpmanual.io/usage/models/#basic-model-usage)
- [ ] [Data serialization](https://pydantic-docs.helpmanual.io/usage/exporting_models/) - `.dict()` and `.json()`
- [ ] [JSON Schema](https://pydantic-docs.helpmanual.io/usage/schema/)
- [X] [Dataclasses](https://pydantic-docs.helpmanual.io/usage/dataclasses/)
- [ ] [Model Config](https://pydantic-docs.helpmanual.io/usage/model_config/)
- [ ] [Field Types](https://pydantic-docs.helpmanual.io/usage/types/) - adding or changing a particular data type
- [ ] [Function validation decorator](https://pydantic-docs.helpmanual.io/usage/validation_decorator/)
- [ ] [Generic Models](https://pydantic-docs.helpmanual.io/usage/models/#generic-models)
- [ ] [Other Model behaviour](https://pydantic-docs.helpmanual.io/usage/models/) - `construct()`, pickling, private attributes, ORM mode
- [ ] [Plugins](https://pydantic-docs.helpmanual.io/) and integration with other tools - mypy, FastAPI, python-devtools, Hypothesis, VS Code, PyCharm, etc.

**Comment by and3rson** [comment]
Possibly related: #2424

**Comment by and3rson** [comment]
Update: this is solved by using `pydantic.dataclasses.dataclass` instead. Still, the error is very confusing.

**Comment by PrettyWood** [comment]
@and3rson This will be fixed in 1.10.3

**Comment by samuelcolvin** [comment]
Closed via #4878
